### PR TITLE
Group metrics generated by `DeviceStatsMonitor` for better visualization

### LINKED ISSUE #11176: Scalars generated by DeviceStatsMonitor should be grouped in tensorboard
### Motivation

Currently, all scalars generated by DeviceStatsMonitor are directly listed in scalar pages of tensorboard.  Half of them start with prefix of "on_train_batch_start", others with "on_train_batch_end". All the keys are joined using dots. Something like:
```
on_train_batch_start.active.all.allocated
on_train_batch_start.active.all.current
on_train_batch_start.active.all.freed
```
In such format, tensorboard will show each scalar value in a seperate entry. This could pollute the scalars tab seriously since there are hundreds of items. 

### Advice
To group the scalars together, one solution could be simply replace the first dot with a slash. So the new names of scalars becomes:

```
on_train_batch_start/active.all.allocated
on_train_batch_start/active.all.current
on_train_batch_start/active.all.freed
```
This time, all scalars are organized into two groups and only two top entries (on_train_batch_start/on_train_batch_end) will be shown in scalars tab of tensorboard.

### Solution

Noticed that the prefix "on_train_batch_start" and "on_train_batch_end" are added by helper function prefix_metrics_keys.

https://github.com/PyTorchLightning/pytorch-lightning/blob/b1baf460d900a1e9982823b4428a3cd108a59d06/pytorch_lightning/callbacks/device_stats_monitor.py#L83

```
def prefix_metrics_keys(metrics_dict: Dict[str, float], prefix: str) -> Dict[str, float]:
    return {prefix + "." + k: v for k, v in metrics_dict.items()}
```
Maybe we can make a minor change like following:
```
def prefix_metrics_keys(metrics_dict: Dict[str, float], prefix: str) -> Dict[str, float]:
    return {prefix + "/" + k: v for k, v in metrics_dict.items()}   # "." -> "/"
```
In this way, minimum code changes are made.

I am not familiar with W&B or other loggers. This suggestion is only based on tensorboard.

#### Before refactor
Screenshot from 2021-12-21 00-12-47

#### After refactor
Screenshot from 2021-12-21 00-09-34

cc @justusschock @awaelchli @akihironitta @kaushikb11

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #11176

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- N/A Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -175,6 +175,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Renamed `training_type_plugin` file to `strategy` ([#11239](https://github.com/PyTorchLightning/pytorch-lightning/pull/11239))
 
 
+- Changed `DeviceStatsMonitor` to group metrics based on the logger's `group_separator` ([]())
```
Comment (rohitgr7) [inline CHANGELOG.md:178]: chlog is missing :)

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/callbacks/device_stats_monitor.py [hunk 2]
```diff
@@ -54,11 +55,15 @@ def on_train_batch_start(
...
     ) -> None:
+        if not trainer.logger:
+            raise MisconfigurationException("Cannot use DeviceStatsMonitor callback with Trainer that has no logger.")
+
         if not trainer.logger_connector.should_update_logs:
             return
 
         device_stats = trainer.accelerator.get_device_stats(pl_module.device)
-        prefixed_device_stats = prefix_metrics_keys(device_stats, "on_train_batch_start")
+        separator = trainer.logger.group_separator
```
Comment (justusschock) [inline pytorch_lightning/callbacks/device_stats_monitor.py:65]: what happens here in the case of a logger collection where loggers might have different separators?
Does the logger collection define this at all?

Comment (ananthsub) [reply pytorch_lightning/callbacks/device_stats_monitor.py:65]: It doesn't. But I think this is a major gap with the idea of logger collections itself, which is why I filed this issue to simplify this further:

File: pytorch_lightning/callbacks/device_stats_monitor.py [hunk 3]
```diff
@@ -54,11 +55,15 @@ def on_train_batch_start(
         batch_idx: int,
         unused: Optional[int] = 0,
     ) -> None:
+        if not trainer.logger:
+            raise MisconfigurationException("Cannot use DeviceStatsMonitor callback with Trainer that has no logger.")
```
Comment (carmocca) [inline pytorch_lightning/callbacks/device_stats_monitor.py:59]: If you add this, you can remove the `assert trainer.logger is not None`s in this file

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/callbacks/device_stats_monitor.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/device_stats_monitor.py:59]:
```suggestion
        if not trainer.logger:
            raise MisconfigurationException("Cannot use `DeviceStatsMonitor` callback with `Trainer(logger=False)`.")
```

**Review by kaushikb11 — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/callbacks/device_stats_monitor.py [hunk 4]
```diff
@@ -71,14 +76,27 @@ def on_train_batch_end(
...
 
-def prefix_metrics_keys(metrics_dict: Dict[str, float], prefix: str) -> Dict[str, float]:
-    return {prefix + "." + k: v for k, v in metrics_dict.items()}
+def _prefix_metric_keys(metrics_dict: Dict[str, float], prefix: str, separator: str) -> Dict[str, float]:
+    return {prefix + separator + k: v for k, v in metrics_dict.items()}
+
+
+def prefix_metric_keys(metrics_dict: Dict[str, float], prefix: str) -> Dict[str, float]:
+    rank_zero_deprecation(
+        "pytorch_lightning.callbacks.device_stats_monitor.prefix_metrics"
```
Comment (Borda) [inline pytorch_lightning/callbacks/device_stats_monitor.py:98]:
```suggestion
        "`pytorch_lightning.callbacks.device_stats_monitor.prefix_metrics`"
```

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/callbacks/device_stats_monitor.py [hunk 5]
```diff
@@ -54,12 +55,15 @@ def on_train_batch_start(
         batch_idx: int,
         unused: Optional[int] = 0,
     ) -> None:
+        if not trainer.logger:
+            raise MisconfigurationException("Cannot use `DeviceStatsMonitor` callback with `Trainer(logger=False)`.")
```
Comment (daniellepintz) [inline pytorch_lightning/callbacks/device_stats_monitor.py:59]: @ananthsub why did you add two more `MisconfigurationException`s here if we already have one in line 48?

Comment (awaelchli) [reply pytorch_lightning/callbacks/device_stats_monitor.py:59]: probably because of mypy, which does not know that the logger was already checked against None in the previous hook :))
